HW0/P5/P5.txt

```diff
@@ -1 +1,92 @@
+
+Therefore, it still takes 8 seconds to verify the sum.
+
+## Question 4
+
+Again using similar logic to the 3rd question, we need to sum the totals one layer of the tree at a time. For N bags, it may not fit perfectly into a binary tree so an additional layer may need to be added. Since each layer takes one second, the total amount of time is the number of layers in the tree excluding the leaves.
+
+This is equal to ceil( lg(N) ).
+
+## Question 5
+
+Assuming there are an infinite number of employees. In this case, we need to also add a cost to the edges connecting layers of the tree. So it takes 1 second for a layer to communicate the total to the layer above.
+
+Continuing from question 3, it takes 8 seconds to calculate the totals since there are 8 layers to compute so therefore there are 7 communications between layers. We assume there is no communication on the base layer (with 256 leaves) since employees are given two bags each at the beginning.
+
+Therefore, the total time taken is 8 + 7 = 15 seconds.
+
+## Question 6
+
+### 1 bag at a time
+
+Since each cashier only takes one bag from us, they will need to either:
+1. Get a bag from another cashier; or
+2. Get the total of that bag from another cashier.
+
+In either case, the time taken is 1 second so for this problem it does not matter. For simplicity, assume each layer is composed of different cashiers so the cashiers on the base of the tree (at the leaves) each go and obtain 1 bag from us. They then communicate (possibly concurrently since that is allowed) the total of their bag to the cashiers at the level above them.
+
+It takes 256 seconds for all cashiers at the base layer to obtain bags. They then communicate the bag totals to the 128 cashiers on the 7th layer in 1 second. Then, the analysis proceeds as in question 5 with calculation and communication taking 15 seconds. It is possible for cashiers to start calculating their totals as soon as they receive their bags, but we will reach a bottleneck since only after 256 seconds do the cashiers on the base layer all have a bag. Therefore, the total time taken is the same regardless of whether they start calculating straight away or only after all base layer cashiers have bags.
+
+The total time taken is 256 + 1 + 8 + 7 = 272.
+
+### 2 bags at a time
+
+On the 7th layer, there are 128 cashiers who each require two bags. This will take 128 seconds if we can only hand bags out at a rate of 2 bags per second (per cashier). Whilst cashiers could get started calculating totals of the upper layers while we are still handing out bags, there is going to be a bottle neck since only after 129 seconds can the 7th layer be entirely completed.
+
+So, including communication time, it takes 128 + 8 + 7 = 143 seconds if we hand each cashier 2 bags at a time.
```

Comment:
The cashiers don't have to wait for all the other cashiers to get a bag before they can start adding/communicating. I don't know if my answers are optimal, but I got 258 and 132 seconds respectively writing a script to do it. Here's my output for 8 bags handing 1 out at a time:

Second 1
Empl:1 
Bags:1

Second 2
Employee 1 communicates to 2
Empl:2 
Bags:2

Second 3
Employee 2 sums
Empl:2 3 
Bags:1 1

Second 4
Employee 2 communicates to 4
Empl:3 4 
Bags:1 2

Second 5
Employee 3 communicates to 5
Employee 4 sums
Empl:4 5 
Bags:1 2

Second 6
Employee 4 communicates to 6
Employee 5 sums
Empl:5 6 
Bags:1 2

Second 7
Employee 5 communicates to 7
Employee 6 sums
Empl:6 7 
Bags:1 2

Second 8
Employee 6 communicates to 8
Employee 7 sums
Empl:7 8 
Bags:1 2

Second 9
Employee 7 communicates to 8
Employee 8 sums
Empl:8 
Bags:2

Second 10
Employee 8 sums
Empl:8 
Bags:1

Completed in 10 seconds

(Sorry for the long comment LOL)